File: dong/go/p56_merge_intervals.go
package main

import (
	"sort"
)

type intss [][]int

func (arr intss) Less(i, j int) bool {
	return arr[i][0] <= arr[j][0]
}

func (arr intss) Swap(i, j int) {
	arr[i], arr[j] = arr[j], arr[i]
}

func (arr intss) Len() int {
	return len(arr)
}

func merge(intervals [][]int) [][]int {
	if len(intervals) <= 1 {
		return intervals
	}
	result := make([][]int, 0)
	var intsP intss = intervals
	sort.Sort(intsP)

	left := intsP[0][0]
	right := intsP[0][1]
	count := 0
	for i := 1; i < len(intervals); i++ {
		if intervals[i][0] > right {
			intervals[count][0] = left
			intervals[count][1] = right
			count++
			left = intervals[i][0]
			right = intervals[i][1]
		} else if intervals[i][1] > right {
			right = intervals[i][1]
		}
	}
	intervals[count][0] = left
	intervals[count][1] = right
	count++

	for i := 0; i < count; i++ {
		result = append(result, intervals[i])
	}

	return result
}

File: dong/java/src/main/java/P56MergeIntervals.java
import java.util.Arrays;
import java.util.Comparator;

/**
 * 给出一个区间的集合，请合并所有重叠的区间。
 *
 * @author Lam
 * @ClassName P56MergeIntervals
 * @date 2020/1/12
 */
public class P56MergeIntervals {
    public int[][] merge(int[][] intervals) {
        if (intervals.length == 0) {
            return new int[0][];
        }
        Arrays.sort(intervals, Comparator.comparingInt(x -> x[0]));
        int left = intervals[0][0];
        int right = intervals[0][1];
        int count = 0;
        for (int i = 1; i < intervals.length; i++) {
            if (intervals[i][0] > right) {
                //不需要合并
                intervals[count][0] = left;
                intervals[count++][1] = right;
                left = intervals[i][0];
                right = intervals[i][1];
            } else if (intervals[i][1] > right) {
                //需要合并
                right = intervals[i][1];
            }
        }
        intervals[count][0] = left;
        intervals[count++][1] = right;
        int[][] result = new int[count][2];
        for (int i = 0; i < count; i++) {
            result[i] = intervals[i];
        }
        return result;
    }
}
